spring-kafka/src/main/java/org/springframework/kafka/listener/FailedRecordProcessor.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import org.springframework.classify.BinaryExceptionClassifier;
+import org.springframework.core.log.LogAccessor;
+import org.springframework.kafka.support.serializer.DeserializationException;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.converter.MessageConversionException;
+import org.springframework.messaging.handler.invocation.MethodArgumentResolutionException;
+import org.springframework.util.Assert;
+import org.springframework.util.backoff.BackOff;
+import org.springframework.util.backoff.FixedBackOff;
+
+/**
+ * Common super class for classes that deal with failing to consume a consumer record.
+ *
+ * @author Gary Russell
+ * @since 2.3.1
+ *
+ */
+public abstract class FailedRecordProcessor {
+
+	private static final BiPredicate<ConsumerRecord<?, ?>, Exception> ALWAYS_SKIP_PREDICATE = (r, e) -> true;
+
+	private static final BiPredicate<ConsumerRecord<?, ?>, Exception> NEVER_SKIP_PREDICATE = (r, e) -> false;
+
+	protected static final LogAccessor LOGGER =
+			new LogAccessor(LogFactory.getLog(SeekToCurrentErrorHandler.class)); // NOSONAR
+
+	private final FailedRecordTracker failureTracker;
+
+	private BinaryExceptionClassifier classifier;
+
+	private boolean commitRecovered;
+
+	protected FailedRecordProcessor(@Nullable BiConsumer<ConsumerRecord<?, ?>, Exception> recoverer, BackOff backOff) {
+		this.failureTracker = new FailedRecordTracker(recoverer, backOff, LOGGER);
+		this.classifier = configureDefaultClassifier();
+	}
+
+	/**
+	 * TODO: remove when the deprecated dependent CTORs are removed.
+	 * @param recoverer the recoverer.
+	 * @param maxFailures the max failures.
+	 */
+	FailedRecordProcessor(@Nullable BiConsumer<ConsumerRecord<?, ?>, Exception> recoverer, int maxFailures) {
```

Comment:
Why do we need to this ctor here if we can just do `FixedBackOff` wrapping in that deprecated ctor in the `DefaultAfterRollbackProcessor`?

Reply:
Because we'd have to duplicate the code in both subclasses.

Reply:
OK. We can live with that. Looks like the matter of personal preferences. 😄